### DISCUSSION TITLE
Efficient configuration to allow data exchange

### DISCUSSION THREAD

**Discussion by youuwang** [GMLC-TDC/HELICS#2689] [Q&A]
Hi, I am recently new to HELICS and I am trying to use it to do co-simulations based on mechanical software (e.g., ABAQUS) and hygrothermal software (e.g., Delphin). Generally, the idea is quite simple. I should have the moisture field published by Delphin at each time step and this should be subscribed by ABAQUS and influence the calculation in ABAQUS. I have a look to the fundamental example on GitHub about battery and charger, but I am still a little confused about how to configure the different federates. For example, in the main loop in Battery.py, helicsFederateRequestTime() is used to synchronize the solvers and it seems that the solver needs to be called in each iteration. But in my case described above, I need to call ABAQUS solver many times, which means that I need to submit a new job for each time step and this is quite inefficient (it will deplete my licenses for ABAQUS).

Could I just define the entire simulation and submit the job once? What is in my mind is: during the simulation, following the synchronization defined by helicsFederateRequestTime(), ABAQUS will just wait for the update of the subscription and apply the new moisture field accordingly for each time step. In this way, I do not need to submit one job per time step. So, is HELICS working like this? Thanks a lot!

**Comment by phlptp** [comment]
HELICS is a library,  it does what you tell it to do, so you can have it work however you like.  If it is really one way communication then you very well could define the entire simulation and submit it once.   The Delphin software would essentially run to completion and publish all the data through its interfaces.   then ABAQUS could do what you want with the data even step through time to collect it all then do the follow-on calculations in a single chunk.   Lots of different ways people can and have used HELICS.

**Reply by youuwang** [reply]
Hi, thanks for the reply! Yes, you are totally correct if it is only a one way communication. I can first finish Delphin, collect the moisture fields at different time steps, and then apply them to different time steps in ABAQUS. But I actually want to prepare for the more complicated co-simulations in the future and I just use this simple case as an example to see whether I can achieve the communication step by step.

So, generally what confuses me is that it seems that I still need to call the solver each time step, as shown in the battery.py and charger.py in the battery-charger example, which means that I need to submit many jobs in ABAQUS. I am wondering how I should adapt the Battery.py file for ABAQUS, for example, to just set up the model and submit the job once, while HELICS could do synchronize the two solvers for data exchange.

**Reply by phlptp** [reply]
I know nothing about ABAQUS or how it works.   But the question comes down to, can you access and submit new data from a running job in ABAQUS,  this can be via callbacks, or remote calls, or really anything as long as some code you wrote is executing during a single ABAQUS job.   If you can do that then you can publish and subscribe to HELICS data in that code.    There are a variety of options for asynchronous operation as well to make HELICS non-blocking if necessary.   

If ABAQUS is a batch processing system (submit data get results), then if you want feedback not sure there are many other options.   Depending on the error tolerances and rate of change you can have different timesteps in each of the simulators to minimize the number of calls.    Others have made approximate models for the faster iterations and only do the expensive calls occasionally when required.  There are lots of games that can be played to minimize solver calls if needed.

**Reply by trevorhardy** [reply]
> So, generally what confuses me is that it seems that I still need to call the solver each time step, as shown in the battery.py and charger.py in the battery-charger example, which means that I need to submit many jobs in ABAQUS. I am wondering how I should adapt the Battery.py file for ABAQUS, for example, to just set up the model and submit the job once, while HELICS could do synchronize the two solvers for data exchange.

I also don't know ABAQUS but it sounds like it doesn't have a sense of time associate with it, perhaps? We've worked with a few solver tools that operate in this manner and have found that wrapping them and calling their associate APIs along with HELICS APIs can give them a sense of time and allow them to fully participate in the co-simulation. If you'd like to discuss this more, we have [office hours this Thursday from 11am to noon PT ](https://helics.org/HELICSOfficeHours.ics)where you could jump on and talk this over more thoroughly.

**Reply by youuwang** [reply]
Hi, thanks for the info! So the office hour is 20:00 for me since I am in Switzerland? As I notice in the .ics file the start time is 20230112T190000Z.

**Comment by trevorhardy** [comment]
11am PDT is 19:00 in Zurich for the next office hours.

And, FYI, the next office hours after this one on the 31st is on Thursday November 14 and will be after the US has its autumnal time change which will move the meeting time for you. (It looks like you just did your time change this past weekend).
